sklearn/manifold/spectral_embedding_.py

```diff
@@ -42,23 +42,23 @@ def _graph_connected_component(graph, node_id):
         belonging to the largest connected components of the given query
         node
     """
-    connected_components_matrix = np.zeros(
-        shape=(graph.shape[0]), dtype=np.bool)
-    nodes_to_explore = np.zeros(shape=(graph.shape[0]), dtype=np.bool)
-    nodes_to_explore[node_id] = True
     n_node = graph.shape[0]
-    nodes_to_add = np.zeros(shape=(graph.shape[0]), dtype=np.bool)
-    for i in range(n_node):
-        nodes_to_add.fill(False)
-        for i in np.where(nodes_to_explore)[0]:
-            nodes_to_add = np.logical_or(nodes_to_add, graph[i] != 0)
-        connected_components_matrix = np.logical_or(
-            connected_components_matrix, nodes_to_explore)
-        if not nodes_to_add.any():
+    connected_nodes = np.zeros(
+        shape=(n_node), dtype=np.bool)
+    nodes_to_explore = np.zeros(shape=(n_node), dtype=np.bool)
+    nodes_to_explore[node_id] = True
+    for _ in range(n_node):
+        last_num_component = connected_nodes.sum()
+        np.logical_or(connected_nodes,
+                      nodes_to_explore,
+                      out=connected_nodes)
+        indices = np.where(nodes_to_explore)[0]
+        nodes_to_explore.fill(False)
+        for i in indices:
+            np.logical_or(nodes_to_explore, graph[i], out=nodes_to_explore)
+        if last_num_component >= connected_nodes.sum():
```

Comment:
I believe that this check for the exit condition can be moved up, just after the update of `connected_nodes`